ext/drwrap/drwrap.c

```diff
@@ -2020,8 +2020,21 @@ drwrap_in_callee(void *arg1, reg_t xsp _IF_NOT_X86(reg_t lr))
 
     NOTIFY(2, "%s: level %d function " PFX "\n", __FUNCTION__, pt->wrap_level + 1, pc);
 
-    drwrap_context_init(drcontext, &wrapcxt, pc, &mc, DRWRAP_WHERE_PRE_FUNC,
-                        IF_X86_ELSE(get_retaddr_from_stack(xsp), (app_pc)lr));
+    app_pc retaddr = IF_X86_ELSE(get_retaddr_from_stack(xsp), (app_pc)lr);
+#ifdef X86
+    if (TEST(DRWRAP_REPLACE_RETADDR, global_flags)) {
+        /* In case of a tailcall for X86, the return address has already been replaced by
+         * the sentinel in the stack, we need to retrieve the return address from the
+         * outer level.
+         */
+        if (retaddr == (app_pc)replace_retaddr_sentinel) {
+            retaddr = pt->retaddr[pt->wrap_level];
```

Comment:
wrap_level starts out -1.  ASSERT that it is >= 0.

Reply:
done.